debezium-core/src/main/java/io/debezium/util/ColumnUtils.java

```diff
@@ -37,7 +42,27 @@ public static ColumnArray toArray(ResultSet resultSet, Table table) throws SQLEx
         Column[] columns = new Column[metaData.getColumnCount()];
         int greatestColumnPosition = 0;
         for (int i = 0; i < columns.length; i++) {
-            columns[i] = table.columnWithName(metaData.getColumnName(i + 1));
+            final String columnName = metaData.getColumnName(i + 1);
+            columns[i] = table.columnWithName(columnName);
+            if (columns[i] == null) {
+                // This situation can happen when SQL Server and Db2 schema is changed before
+                // an incremental snapshot is started and no event with the new schema has been
+                // streamed yet.
+                // This warning will help to identify the issue in case of a support request.
+
+                final String[] resultSetColumns = new String[metaData.getColumnCount()];
+                for (int j = 0; j < metaData.getColumnCount(); j++) {
+                    resultSetColumns[j] = metaData.getColumnName(j + 1);
+                }
+                LOGGER.warn(
+                        "Column '{}' not found in result set '{}' for table '{}', {}. This might be caused by DBZ-4350",
```

Comment:
It seems that this check doesn't cover column type changes or a case when a column gets dropped. I think, what you need here is to make sure that the `Table` and `ResultSetMetaData` have the same schema and https://github.com/debezium/debezium/pull/2853 does exactly that

Reply:
@jpechane, it seems this comment still needs addressing? Or has it been taken care of already elsewhere?

Reply:
Btw. why creating (but not throwing) an IAE only for logging and then continuing?

Reply:
That's the question - whether we want to stop it right away or not. Currently it is not and I just wanted to capture how we got there.
Regarding the other comment, the code is aimed to fix NPE only, it is not intended to handle all cases.
Kate refers to PR that is already fixed and originally thought that this somehow builts upon it but as I already mentioned this is only bandaid and if we take a great effort just to catch all cases when the user has triggered an unsupportedoperation then we can at the end raather spend all this effort on fixing it. The code need not be there at all, the SQL Server will still be unsupported.

Reply:
> The code need not be there at all, the SQL Server will still be unsupported.

I don't understand that part.

> That's the question - whether we want to stop it right away or not.

Yes, right. But why create an exception in the "log only" approach at all?

So taking a step back, the priorities we should consider IMO are these:

* Consistency trumps everything; i.e. we should not emit snapshot events with an incorrect schema. Instead, if we cannot automatically handle this situation by refreshing the schema, the snapshot should fail with an indicative message, allowing the user to take the required action (e.g. do a schema migration for the SQL Server / Db2 connectors) before re-triggering the snapshot
* Automate things as much as possible (this can be done in follow-up PRs)
* If there's known limitations (e.g. schema changes in parallel to incremental snapshots not supported for some connector), they should be clearly called out in the docs

Reply:
@gunnarmorling The docs commit has just been pushed.

Reply:
@gunnarmorling Excaption thrown.